doc/kernel/object_cores/index.rst

```diff
@@ -0,0 +1,234 @@
+Object cores provide a uniform means to retrieve that information via object
+core statistics. When enabled, the object type contains a pointer to a
+statistics descriptor that defines the various operations that have been
+enabled for interfacing with the object's statistics. Additionally, the object
+core contains a pointer to the "raw" statistical information associated with
+that object. Raw data is the raw, unmanipulated data associated with the
+statistics. Queried data may be "raw", but it may also have been manipulated in
+some way by calculation (such as determining an average).
+
+Object core statistics have been integrated into the following kernel objects:
+The following table indicates both what objects have been integrated into the
+object core statistics as well as the structures used for both "raw" and
+"queried" data.
+
+=====================  ============================== ==============================
+Object                 Raw Data Type                  Query Data Type
+=====================  ============================== ==============================
+struct mem_slab        struct mem_slab_info            struct sys_memory_stats
+struct sys_mem_blocks  struct sys_mem_blocks_info      struct sys_memory_stats
+struct k_thread        struct k_cycle_stats            struct k_thread_runtime_stats
+struct _cpu            struct k_cycle_stats            struct k_thread_runtime_stats
+struct z_kernel        struct k_cycle_stats[num CPUs]  struct k_thread_runtime_stats
+=====================  ============================== ==============================
+
+Implementation
+**************
+
+Defining a New Object Type
+==========================
+
+An object type is defined using a global variable of type
+:c:struct:`k_obj_type`. It must be initialized before any objects of that type
+are initialized. The following code shows how a new object type can be
+initialized for use with object cores and object core statistics.
+
+.. code-block:: c
+
+    /* Unique object type ID */
+
+    #define K_OBJ_TYPE_MY_NEW_TYPE  K_OBJ_TYPE_ID_GEN("UNIQ")
+    struct k_obj_type  my_obj_type;
+
+    struct my_obj_type_raw_info {
+        ...
+    };
+
+    struct my_obj_type_query_stats {
+        ...
+    };
+
+    struct my_new_obj {
+        ...
+        struct k_obj_core obj_core;
+        struct my_obj_type_raw_info  info;
+    };
+
+    struct k_obj_core_stats_desc my_obj_type_stats_desc = {
+        .raw_size = sizeof(struct my_obj_type_raw_stats),
+        .query_size = sizeof(struct my_obj_type_query_stats),
+        .raw = my_obj_type_stats_raw,
+        .query = my_obj_type_stats_query,
+        .reset = my_obj_type_stats_reset,
+        .disable = NULL,    /* Stats gathering is always on */
+        .enable = NULL,     /* Stats gathering is always on */
+    };
+
+    void my_obj_type_init(void)
+    {
+        z_obj_type_init(&my_obj_type, K_OBJ_TYPE_MY_NEW_TYPE,
+                        offsetof(struct my_new_type, obj_core);
```

Comment:
Shouldn't this be `struct my_new_obj` instead of `struct my_new_type`? I.e. what you have on line 93.

Reply:
@peter-mitsis would it make things easier to add an actual code sample demonstrating this? This would help keep the code free of typos, and you wwould still be embed to embed snippets of it in this documentation page using `.. literalinclude:` if you wanted to.

Reply:
@jhedberg ... you are correct. I will be fixing this shortly.

@kartben ... I am working on a sample project for this.